File: aws_interface/sdk/nodejs/test/storageTest.js
// databaseTest.js

const Client = require('../aws_interface.js');
const assert = require("assert");
const fs = require('fs');

var client = new Client.Client();
const test_email = "ttest@gmail.com";
const test_password = "ttestpassword";
const read_groups = ['user', 'admin'];
const write_groups = ['user', 'admin'];
const file_name = 'test.txt'
var file_id;
var file_b64;


describe('Storage Test', function() {

    //Create a textfile for testing.
    let data = "This is test file.";
    fs.writeFile('test.txt', data, (err) => {
        if (err) throw err;
    });

    // Upload a file with [file_name] with read permissions to [read_groups], write permissions to [write_groups]
    // Get [file_id] when uploading.
    // Download and delete the file with [file_id].
    // Wait for the upload and download to finish by async and await on response.
    it('test storageUploadFile,storageDownloadFile, storageDeleteFile', function(done) {
        client.authLogin(test_email, test_password, function (body) {
            client.storageUploadFile(file_name, read_groups, write_groups, async function (response) {
                file_id = await response['file_id'];
                assert.ok(response['file_id']);

                client.storageDownloadFile(file_id, async function (response) {
                    const rr = await response;
                    assert.ok(response);
                });

                client.storageDeleteFile(file_id, function (response) {
                    assert.ok(response['success']);
                });

            });
            done();
        });
    });
/*
    it('test storageDownloadB64Chunk, storageDeleteB64', function (done) {
        client.authLogin(test_email, test_password, function (body) {
            file_id = "N7PxzRdEK3nhE3GM53CAnb";

            client.storageDownloadB64Chunk(null, async function (response) {
                    file_b64 = await response['file_b64'];
                    console.log(response);
                    //assert.ok(response['file_64']);
                });
                client.storageDeleteB64(file_id, function (response) {
                    assert.ok(response['success']);
                    console.log(response);
                });

            done();
        });
    });
    it('test storageUploadB64Chunk', function (done) {
        client.authLogin(test_email, test_password, function (body) {

            client.storageDeleteB64(file_id, function (response) {
                assert.ok(response['success']);
            });
        });
        done();
    });
 */
});

File: aws_interface/sdk/java/source/src/test/java/StorageTest.java
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import org.junit.jupiter.api.*;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.assertEquals;
import java.util.concurrent.Callable;
import static org.awaitility.Awaitility.*;
import static java.util.concurrent.TimeUnit.SECONDS;

public class StorageTest {
	AWSInterface client = new AWSInterface("https://e3t7xfb981.execute-api.ap-northeast-2.amazonaws.com/prod_aws_interface/ipxsNKvdpXUgzAMqg2JgdM");
	String test_email = new String("ttest@gmail.com");
	String test_password = new String("ttestpassword");
	static ArrayList<String> readGroups = new ArrayList<>();
	static ArrayList<String> writeGroups = new ArrayList<>();
	
	static String file_dir = System.getProperty("user.dir");
	static String file_path = Paths.get(file_dir, "test.txt").toString();
	static String file_id ="G9vDDvTz8AT54TS4qwsXKJ";
	static File file;
	boolean received = false;
	boolean received2 = false;
	
	//Declare Callable object for check if the response is received;
	private Callable<Boolean> responseReceived(){
		return new Callable<Boolean>() {
			public Boolean call() throws Exception{
				return received;
			}
		};
	}
	
	
	//Declare Callable object for check if the response is received;
	private Callable<Boolean> responseReceived2(){
		return new Callable<Boolean>() {
			public Boolean call() throws Exception{
				return received2;
			}
		};
	}
	
	@BeforeAll
	static public void writeFile() {
		try {
			file = new File(file_path);
			FileWriter fw = new FileWriter(file, false);
			fw.write("this is test text");
			fw.flush();
			fw.close();
		}
		catch (Exception e) {
			e.printStackTrace();
		}
	}

	@BeforeAll
	static public void setValues() {
		 readGroups.add("user");
		 writeGroups.add("user");

	}
	// Before each testcase, initialize received flag with false to use awaitility for async testing
	@BeforeEach
	public void initialiize() {
		received = false;
		received2 = false;
	}

	@Test
    public void storageTest(){
    	client.authLogin(test_email, test_password,(rr, hE)->{
		try {
			//UploadFile with [readGroups], [writeGroups] and save file_id to [file_id]
	        	client.storageUploadFile(file, (response, hasError)->{
	        		file_id = response.get("file_id").getAsString();
	        		assertNotNull(response.get("file_id").getAsString());
	        		System.out.println("upload "+ response);
	    	        received2 = true;
            });
      }
		catch (Exception e){
			e.printStackTrace();
        }
	});
    	//Wait for the upload to finish at most 20 seconds
    	await().atMost(20, SECONDS).until(responseReceived2());
    	received2 = false;

    	//Download file with file_id
    	client.storageDownloadFileByte(file_id, (response, hasError)->{
            System.out.println("download "+response);
            //assertTrue(response.get("success").getAsBoolean());
            received2 = true;
        }, (filebyte)->{});
    	//Wait for the download to finish at most 20 seconds
    	await().atMost(20, SECONDS).until(responseReceived2());
    	
    	//Delet the file with file_id
    	client.storageDeleteFile(file_id,(response, hasError)->{
            System.out.println("delete "+response);
            //assertTrue(response.get("success").getAsBoolean());
            received = true;
        });
    	//Wait for all the processes at most 30 seconds
        await().atMost(30, SECONDS).until(responseReceived());
    }
}


